crates/bevy_animation/src/entity_path.rs

```diff
@@ -0,0 +1,153 @@
+//! Entity paths for referring to bones.
+
+use std::{
+    fmt::{self, Debug, Formatter, Write},
+    hash::{Hash, Hasher},
+    sync::{Arc, Mutex, OnceLock, Weak},
+};
+
+use bevy_core::Name;
+use bevy_reflect::Reflect;
+use bevy_utils::prelude::default;
+use weak_table::WeakHashSet;
+
+static ENTITY_PATH_STORE: OnceLock<Mutex<EntityPathStore>> = OnceLock::new();
+
+/// Path to an entity, with [`Name`]s. Each entity in a path must have a name.
+#[derive(Clone, Reflect)]
+#[reflect_value]
+pub struct EntityPath(Arc<EntityPathNode>);
+
+#[derive(PartialEq, Eq, Hash)]
+struct EntityPathNode {
+    name: Name,
+    parent: Option<EntityPath>,
+}
+
+// This could use a `RwLock`, but we actually never read from this, so a mutex
+// is actually slightly more efficient!
+#[derive(Default)]
+struct EntityPathStore(WeakHashSet<Weak<EntityPathNode>>);
+
+pub struct EntityPathIter<'a>(Option<&'a EntityPath>);
+
+impl EntityPathStore {
+    fn create_path(&mut self, node: EntityPathNode) -> EntityPath {
+        match self.0.get(&node) {
+            Some(node) => EntityPath(node),
+            None => {
+                let node = Arc::new(node);
+                self.0.insert(node.clone());
+                EntityPath(node)
+            }
+        }
+    }
+}
+
+impl EntityPath {
+    pub fn from_name(name: Name) -> EntityPath {
+        ENTITY_PATH_STORE
+            .get_or_init(|| default())
+            .lock()
+            .unwrap()
+            .create_path(EntityPathNode { name, parent: None })
+    }
+
+    pub fn from_names(names: &[Name]) -> EntityPath {
+        let mut store = ENTITY_PATH_STORE.get_or_init(|| default()).lock().unwrap();
+
+        let mut names = names.iter();
+        let root_name = names
+            .next()
+            .expect("Entity path must have at least one name in it");
+
+        let mut path = store.create_path(EntityPathNode {
+            name: root_name.clone(),
+            parent: None,
+        });
+        for name in names {
+            path = store.create_path(EntityPathNode {
+                name: name.clone(),
+                parent: Some(path),
+            });
+        }
+
+        path
+    }
+
+    pub fn extend(&self, name: Name) -> EntityPath {
+        ENTITY_PATH_STORE
+            .get_or_init(|| default())
+            .lock()
+            .unwrap()
+            .create_path(EntityPathNode {
+                name,
+                parent: Some(self.clone()),
+            })
+    }
+
+    pub fn iter(&self) -> EntityPathIter {
+        EntityPathIter(Some(self))
+    }
+
+    pub fn root(&self) -> &Name {
+        &self.iter().last().unwrap().0.name
+    }
```

Comment:
It's kind of unfortunate that common operations like `root` and (forward) `iter` have to be less efficient due to the way the linked list is constructed. Is there a reason it cannot be created reversed (or the whole path be interned together)?